category/applicative.lean

```diff
@@ -0,0 +1,153 @@
+/-
+Copyright (c) 2017 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+
+Instances for identity and composition functors
+-/
+
+import category.functor
+
+universe variables u v w u' v' w'
+
+section lemmas
+
+open function applicative
+
+variables {α β γ : Type u}
+variables {f : Type u → Type v}
+variables [applicative f] [is_lawful_applicative f]
+variables (g : β → γ)
+
+lemma pure_seq_pure {α β : Type u} (g : α → β) (x : α) :
+  pure g <*> (pure x : f α) = pure (g x) :=
+by simp
+
+open function
+
+lemma applicative.map_seq {β γ σ : Type u} (h : γ → σ) (x : f (β → γ)) (y : f β) :
+  h <$> (x <*> y) = (comp h <$> x) <*> y :=
+by rw [← pure_seq_eq_map,← pure_seq_eq_map,
+       seq_assoc,map_pure]
+
+lemma applicative.seq_map {β γ σ : Type u} (h : σ → β) (x : f (β → γ)) (y : f σ) :
+  x <*> (h <$> y) = (flip comp h) <$> x <*> y :=
+begin
+  rw [← pure_seq_eq_map,← pure_seq_eq_map,
+      seq_assoc,
+      seq_pure,
+      pure_seq_eq_map,
+      ← comp_map] ,
+  refl
+end
+
+open applicative functor
+
+attribute [norm] seq_assoc pure_seq_eq_map map_pure seq_map map_seq
+
+lemma applicative.map_seq_map {α β γ σ : Type u} (g : α → β → γ) (h : σ → β) (x : f α) (y : f σ) :
+  (g <$> x) <*> (h <$> y) = (flip comp h ∘ g) <$> x <*> y :=
+by simp with norm
+
+lemma applicative.map_seq_assoc
+  (x : f (α → β)) (y : f α) :
+  @functor.map f _ _ _ g (x <*> y) = comp g <$> x <*> y :=
+by rw [← pure_seq_eq_map,
+       seq_assoc, map_pure,
+       pure_seq_eq_map]
+
+lemma applicative.seq_map_comm
+  (x : f (γ → α)) (y : f β) :
+  x <*> g <$> y = flip comp g <$> x <*> y :=
+begin
+  rw [← pure_seq_eq_map _ y, seq_assoc, seq_pure, ← comp_map],
+  refl,
+end
+
+end lemmas
+
+instance applicative_id : applicative id :=
+{ pure := @id,
+  seq := λ α β f, f }
+instance lawful_applicative_id : is_lawful_applicative id :=
+by refine { .. } ; intros ; refl
+
+namespace compose
+
+open function
+
+variables {f : Type u → Type u'} {g : Type v → Type u}
+
+variables [applicative f] [applicative g]
+
+def seq  {α β : Type v} : compose f g (α → β) → compose f g α → compose f g β
+  | ⟨ h ⟩ ⟨ x ⟩ := ⟨ has_seq.seq <$> h <*> x ⟩
+
+instance : has_pure (compose f g) :=
```

Comment:
I don't think any of these instances work - lean has a tendency to misbehave when doing type class inference with `compose` since it eagerly inserts an identity function on one side. Have you tested this?

Reply:
Yes, that was quite the headache so I built a structure around `compose`:

https://github.com/cipher1024/mathlib/blob/traversable/category/functor.lean#L35-L36